src/crimson/os/seastore/async_cleaner.cc

```diff
@@ -905,23 +905,11 @@ AsyncCleaner::retrieve_backref_mappings(
   paddr_t start_paddr,
   paddr_t end_paddr)
 {
-  return seastar::do_with(
-    backref_pin_list_t(),
-    [this, start_paddr, end_paddr](auto &pin_list) {
-    return repeat_eagain([this, start_paddr, end_paddr, &pin_list] {
-      return ecb->with_transaction_intr(
-	Transaction::src_t::READ,
-	"get_backref_mappings",
-	[this, start_paddr, end_paddr](auto &t) {
-	return backref_manager.get_mappings(
-	  t, start_paddr, end_paddr
-	);
-      }).safe_then([&pin_list](auto&& list) {
-	pin_list = std::move(list);
-      });
-    }).safe_then([&pin_list] {
-      return seastar::make_ready_future<backref_pin_list_t>(std::move(pin_list));
-    });
+  return ecb->with_transaction_weak(
+      "backref_get_mappings",
```

Comment:
Can I ask a question about why did you change the original transaction to weak transaction?

Reply:
There is a restriction from the current backref-tree implementation, that unlike lba tree, it doesn't support tree-read during tree-updates or concurrent tree-updates with parallel transactions. This is why our gc transactions are called in serial, and why I changed this READ transaction to weak for readability. Probably I need to add more comments here.

I'm not sure about the specific gap in the backref-tree to enable parallism, probably @xxhdx1985126 is able to elaborate.

Reply:
The reason why I prevented backref-tree nodes from concurrent accessing was to prevent future developers from merging in-cache backref entries in parallel or updating the backref tree in client io transactions. Because the backref tree's key is `paddr_t`, if we allow updating backref tree in client io transactions or parallel in-cache backref entries merging, in the scenario of single segmented device, all those transactions are probably gonna conflict on the same backref node, which could hurt the performance of seastore very badly.

But, when it comes to the multi-device scenario, I think we can loose that restriction a little bit so that cleaners of different devices can update the backref tree concurrently, because `paddrs` of different devices must exist in whole different sub-trees and, unless the number of devices exceeds the size of backref tree's internal node, those transactions can only conflict on the root of the backref tree. Am I right? @cyx1231st @myoungwon @athanatos

Reply:
> Because the backref tree's key is paddr_t, if we allow updating backref tree in client io transactions or parallel in-cache backref entries merging, in the scenario of single segmented device, all those transactions are probably gonna conflict on the same backref node, which could hurt the performance of seastore very badly.

I agree with the point. In addition, the backref tree is a mirror of the lba tree, so all the logical conflicts should have been addressed by the lba tree already. Further conflicts in the backref tree seem to be unnecessary.

> But, when it comes to the multi-device scenario, I think we can loose that restriction a little bit so that cleaners of different devices can update the backref tree concurrently ...

Is there really any restrictions here? Only the alloc-trim transaction is updating the backref-tree, and seems to me only the primary tier is responsible to do journaling and trimming. IIUC there is no real requirement to allow concurrent updates to the backref tree. @xxhdx1985126

Reply:
> Is there really any restrictions here? Only the alloc-trim transaction is updating the backref-tree, and seems to me only the primary tier is responsible to do journaling and trimming. IIUC there is no real requirement to allow concurrent updates to the backref tree. @xxhdx1985126

Ah, yes! hah

Reply:
Right, so we should not allow concurrent transactions between trim and reclaim right now.

Reply:
Understood.

Reply:
> Right, so we should not allow concurrent transactions between trim and reclaim right now.

Um, actually, I think maybe we can allow this in the multi-device scenario. Because paddr mappings of different devices should be in different subtrees, the parent of which is the backref tree root. So, the alloc-trim transactions of the primary cleaner can only conflict with reclaim transactions of other cleaner when the alloc-trim transactions touch the backref tree root, which might be a relatively low-possibility event. But, I'm not sure whether it's really that low, might need more tests on this issue.

Reply:
> Right, so we should not allow concurrent transactions between trim and reclaim right now.

Yeah

Reply:
> Probably I need to add more comments here.

Comments added.